bundle/src/main/java/com/adobe/acs/commons/mcp/impl/processes/TreeReplication.java

```diff
@@ -0,0 +1,272 @@
+    }
+
+    List<EnumMap<ReportColumns, String>> reportData = Collections.synchronizedList(new ArrayList<>());
+
+    private void record(String path, String action, String description) {
+        EnumMap<ReportColumns, String> row = new EnumMap<>(ReportColumns.class);
+        row.put(ReportColumns.PATH, path);
+        row.put(ReportColumns.ACTION, action);
+        row.put(ReportColumns.DESCRIPTION, description);
+        reportData.add(row);
+    }
+
+    @Override
+    public void storeReport(ProcessInstance instance, ResourceResolver rr) throws RepositoryException, PersistenceException {
+        GenericReport report = new GenericReport();
+        report.setName("Tree Replication " + startingPath);
+        report.setRows(reportData, ReportColumns.class);
+        report.persist(rr, instance.getPath() + "/jcr:content/report");
+    }
+
+    // Should match nt:folder, sling:OrderedFolder, sling:UnorderedFolder, etc
+    public static Boolean isFolderOrPage(Resource res) {
+        String primaryType = String.valueOf(res.getResourceType()).toLowerCase();
+        return (primaryType.contains("folder") || primaryType.equalsIgnoreCase(NameConstants.NT_PAGE));
+    }
+
+    private void activateTreeStructure(ActionManager t) {
+        TreeFilteringResourceVisitor visitor = createFolderPageVisitor();
+        visitor.setResourceVisitorChecked((resource, u) -> {
+            String path = resource.getPath();
+            if (publishFilter.shouldReplicate(resource)) {
+                t.withResolver(rr -> performReplication(t, path));
+            } else {
+                record(path, "Skip", "Skipping folder");
+            }
+        });
+        t.deferredWithResolver(rr -> visitor.accept(rr.getResource(startingPath)));
+    }
+
+    private void deactivateTreeStructure(ActionManager t) {
+        t.deferredWithResolver(rr -> {
+            performAsynchronousReplication(t, startingPath);
+        });
+    }
+
+    private void activateContent(ActionManager t) {
+        TreeFilteringResourceVisitor visitor = createFolderPageVisitor();
+        visitor.setLeafVisitorChecked((resource, u) -> {
+            String path = resource.getPath();
+            if (publishFilter.shouldReplicate(resource)) {
+                t.deferredWithResolver(rr -> performReplication(t, path));
+            } else {
+                record(path, "Skip", "Skipping content");
+            }
+        });
+        t.deferredWithResolver(rr -> visitor.accept(rr.getResource(startingPath)));
+    }
+
+    public AtomicInteger replicationCount = new AtomicInteger();
+
+    private void performReplication(ActionManager t, String path) {
+        int counter = replicationCount.incrementAndGet();
+        if (queueMethod == QueueMethod.USE_MCP_QUEUE
+                || (queueMethod == QueueMethod.MCP_AFTER_10K && counter >= ASYNC_LIMIT)) {
+            performSynchronousReplication(t, path);
+        } else {
+            performAsynchronousReplication(t, path);
```

Comment:
Just to make sure I understand, if someone chooses NOT to use MCP they will use standard async replication which will toss the jobs into the sling job queue as fast this process can walk the tree right?

Reply:
Correct.  And god help them if they do that with 20,000+ nodes.

Reply:
The advantage of using MCP to manage the publish queue is twofold: 1) You can't kill AEM as easily even if you're publishing tens of thousands of things with it.  Standard product queue will seize up between 50k-100k items.  2) You get a progress bar that lets you know how much has been published and how much is left.  Oh, and that nice report at the end.

Reply:
yup - cool - just wanted to make sure I understood. Good default value :) 👍